internal/step/plugin/provider.go

```diff
@@ -885,43 +1027,61 @@ func (r *runningStep) deployStage() (deployer.Plugin, error) {
 	return container, nil
 }
 
+// enableStage returns the result of whether the stage was enabled or not.
+func (r *runningStep) enableStage() (bool, error) {
+	r.lock.Lock()
+	previousStage := string(r.currentStage)
+	r.currentStage = StageIDEnabling
+	enabledInputAvailable := r.enabledInputAvailable
+	r.state = step.RunningStepStateWaitingForInput
+	r.lock.Unlock()
+
+	r.stageChangeHandler.OnStageChange(
+		r,
+		&previousStage,
+		nil,
+		nil,
+		string(StageIDEnabling),
+		enabledInputAvailable,
+		&r.wg,
+	)
+
+	select {
+	case enabled := <-r.enabledInput:
+		return enabled, nil
+	case <-r.ctx.Done():
+		return false, fmt.Errorf("step closed while determining enablement status")
+	}
+}
+
 func (r *runningStep) startStage(container deployer.Plugin) error {
 	r.logger.Debugf("Starting stage for step %s/%s", r.runID, r.pluginStepID)
 	atpClient := atp.NewClientWithLogger(container, r.logger)
 	r.lock.Lock()
-	previousStage := string(r.currentStage)
-	r.currentStage = StageIDStarting
 	inputRecievedEarly := false
 	r.atpClient = atpClient
+	r.lock.Unlock()
 
 	var runInput any
+	var newState step.RunningStepState
 	select {
 	case runInput = <-r.runInput:
 		// Good. It received it immediately.
-		r.state = step.RunningStepStateRunning
+		newState = step.RunningStepStateRunning
 		inputRecievedEarly = true
 	default: // The default makes it not wait.
-		r.state = step.RunningStepStateWaitingForInput
+		newState = step.RunningStepStateWaitingForInput
 	}
```

Comment:
We don't need to set `inputRecievedEarly` under the protection of the lock (and, doing so means that we're holding the lock longer, which is not good).

I suggest declaring `inputRecievedEarly` with a `var` (like `newState`, at line 1066) and setting it to `false` after setting `newState` at line 1073 analogously to how it is set to `true` after setting `newState` at line 1070.

(Also, I think `inputRecievedEarly` is misspelled....)